### PR TITLE
ASoC: soundwire: fix broken BIOS with _ADR workaround

### PR DESCRIPTION
**Pull request by plbossart**
On some HP devices, the BIOS does not advertise the presence of RT711
on link0 and RT1308 on link1. This has been a known issue on reference
platforms, but this problem was not expected to spill into commercial
devices.

Unfortunately the BIOS remained broken and we can't reasonably expect
every Linux user to override the DSDT.

This patch suggests the use of a workaround. Broken BIOS expose the
part 0x701 on all links, which does not correspond to a real device or
configurations. We can however use this information to make existing
codec drivers can probe on this device instead.

BugLink: https://github.com/thesofproject/linux/issues/2700
Signed-off-by: Pierre-Louis Bossart <pierre-louis.bossart@linux.intel.com>

### CODE REVIEW

**Comment by plbossart** [comment]
pushed the wrong version, fixed now.

**Comment by plbossart** [comment]
@bardliao @RanderWang I just pushed a new version that works on CML-U w/ SoundWire, where I aliased rt700-link0 to rt711 and rt700-link1 to rt1308-1. 
Can you take a look? To test on the HP device, we only need to change 0x700 -> 0x701 for TGL. Thanks!

Also we need a rt711-rt1308-mono topology with DMICs for that configuration. Is this something you can prepare (should be similar to TGL-RVP but there's a single RT1308, not 2 on link1).

**Comment by RanderWang** [comment]
> @bardliao @RanderWang I just pushed a new version that works on CML-U w/ SoundWire, where I aliased rt700-link0 to rt711 and rt700-link1 to rt1308-1.
> Can you take a look? To test on the HP device, we only need to change 0x700 -> 0x701 for TGL. Thanks!
> 
> Also we need a rt711-rt1308-mono topology with DMICs for that configuration. Is this something you can prepare (should be similar to TGL-RVP but there's a single RT1308, not 2 on link1).

sure, I will test it on tgl rvp

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 1]
```diff
@@ -305,6 +323,20 @@ static const struct snd_soc_acpi_link_adr tgl_3_in_1_sdca[] = {
...
+static const struct snd_soc_acpi_link_adr tgl_workaround_hp[] = {
+	{
+		.mask = BIT(0),
+		.num_adr = ARRAY_SIZE(workaround_rt711_0_adr),
+		.adr_d = workaround_rt711_0_adr,
+	},
+	{
+		.mask = BIT(1),
+		.num_adr = ARRAY_SIZE(workaround_rt1308_1_single_adr),
+		.adr_d = rt1308_1_single_adr,
```
Comment (RanderWang) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: is it workaround_rt1308_1_single_adr ?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 2]
```diff
@@ -77,6 +86,15 @@ static const struct snd_soc_acpi_adr_device rt1308_1_single_adr[] = {
 	}
 };
 
+static const struct snd_soc_acpi_adr_device workaround_rt1308_1_single_adr[] = {
+	{
+		.adr = 0x000110025D070100,
+		.num_endpoints = 1,
```
Comment (RanderWang) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c:95]: we need  		.adr_override = 0x000120025D130800, /* same as rt1308_1_single_adr */ ?

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c [hunk 3]
```diff
@@ -53,6 +53,15 @@ static const struct snd_soc_acpi_adr_device rt711_1_adr[] = {
 	}
 };
 
+static const struct snd_soc_acpi_adr_device workaround_rt711_0_adr[] = {
+	{
+		.adr = 0x000010025D070100, /* using broken _ADR as mapping to rt711 */
+		.num_endpoints = 1,
```
Comment (RanderWang) [inline sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: override ?

**Comment by RanderWang** [comment]
@plbossart  yes, we need to change 0x700 -> 0x701 (int codec driver) for TGL. With above fix, it can work on TGL RVP.

File: drivers/soundwire/bus_type.c [hunk 4]
```diff
@@ -174,9 +178,15 @@ static void sdw_drv_shutdown(struct device *dev)
...
+
 	drv->driver.bus = &sdw_bus_type;
 
 	if (!drv->probe) {
+		name = drv->name;
+		if (!name)
+			name = drv->driver.name;
+
 		pr_err("driver %s didn't provide SDW probe routine\n",
 		       drv->name);
```
Comment (lyakh) [inline drivers/soundwire/bus_type.c]: didn't you mean to replace `drv->name` here too?

**Review by lyakh — CHANGES_REQUESTED** [review]

File: drivers/soundwire/bus_type.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/bus_type.c]: yep, good catch @lyakh !

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: yes, I didn't touch tgl for now, this is needed as well.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:95]: yes, I didn't touch tgl for now, this is needed as well.

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: yes, I didn't touch tgl for now, this is needed as well.

File: drivers/soundwire/bus_type.c [hunk 5]
```diff
@@ -108,7 +109,10 @@ static int sdw_drv_probe(struct device *dev)
 
 	ret = drv->probe(slave, id);
 	if (ret) {
-		dev_err(dev, "Probe of %s failed: %d\n", drv->name, ret);
+		name = drv->name;
+		if (!name)
+			name = drv->driver.name;
+		dev_err(dev, "Probe of %s failed: %d\n", name, ret);
```
Comment (kv2019i) [inline drivers/soundwire/bus_type.c:115]: Maybe add a macro to look up "sdw_driver_name(sdw_driver *driver)"? I'd guess this is needed in many places. Or maybe just drop the top-level name.

File: sound/soc/codecs/rt711-sdw.c [hunk 6]
```diff
@@ -443,6 +443,13 @@ static int rt711_sdw_probe(struct sdw_slave *slave,
 {
 	struct regmap *sdw_regmap, *regmap;
 
+	/* filter BIOS work-around on some Intel platforms */
```
Comment (kv2019i) [inline sound/soc/codecs/rt711-sdw.c]: I had to read the above comment a few times. This isn't "filtering a workaround", right? Maybe "/* catch faulty BIOS configuration on some Intel platforms */"

File: sound/soc/codecs/rt711-sdw.c [hunk 7]
```diff
@@ -473,6 +480,8 @@ static int rt711_sdw_remove(struct sdw_slave *slave)
 
 static const struct sdw_device_id rt711_id[] = {
 	SDW_SLAVE_ENTRY_EXT(0x025d, 0x711, 0x2, 0, 0),
+	/* fake part to work-around missing _ADR in BIOS */
```
Comment (kv2019i) [inline sound/soc/codecs/rt711-sdw.c]: This is ok, but just raising a note that it isn't immediately obvious "part" is referring to the "part" field of SDW. But yeah, people more familiar with will understand this for sure. Maybe "fake the part field to..." would be ever more clear.

**Review by kv2019i — COMMENTED** [review]
Complicated but in general looks ok. A few notes on readability from a not-so-familiar-with-sdw point of view.

File: drivers/soundwire/bus_type.c (hunk 4 above)
Comment (plbossart) [reply drivers/soundwire/bus_type.c]: fixed now

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 3 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: fixed now

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 2 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c:95]: fixed now

File: sound/soc/intel/common/soc-acpi-intel-tgl-match.c (hunk 1 above)
Comment (plbossart) [reply sound/soc/intel/common/soc-acpi-intel-tgl-match.c]: fixed now

File: sound/soc/codecs/rt711-sdw.c (hunk 7 above)
Comment (plbossart) [reply sound/soc/codecs/rt711-sdw.c]: yes, I can improve it as 'fake part_id'

File: sound/soc/codecs/rt711-sdw.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/codecs/rt711-sdw.c]: it's really 'only allow for BIOS workaround in specific configurations'

File: drivers/soundwire/bus_type.c (hunk 5 above)
Comment (plbossart) [reply drivers/soundwire/bus_type.c:115]: i'll see what Vinod wants when upstreaming, this is a cosmetic detail.

File: sound/soc/codecs/rt711-sdw.c (hunk 6 above)
Comment (plbossart) [reply sound/soc/codecs/rt711-sdw.c]: fixed now

**Comment by plbossart** [comment]
@RanderWang can you give this PR a try with no ACPI initrd override?

I was mistaken, we already have the sof-tpg-rt711-rt1308 which can work. However we are missing the DMIC autodetection, we need to figure out how to add the -2ch or -4ch suffix automagically based on NHLT information. We can't ask users to manually rename topology filenames.

**Comment by RanderWang** [comment]
> @RanderWang can you give this PR a try with no ACPI initrd override?
> 
> I was mistaken, we already have the sof-tpg-rt711-rt1308 which can work. However we are missing the DMIC autodetection, we need to figure out how to add the -2ch or -4ch suffix automagically based on NHLT information. We can't ask users to manually rename topology filenames.

The RT711 on my TGL RVP is a external one and it can't work properly for many bus clash but rt1308 is in good status.
The status of this PR on my TGL:
(1) sound card is registered successfully and all pcm stream are set up correctly
(2) I didn't get any error for playback on RT1308

update: I have removed ACPI override yesterday so I could test with this workaround.

**Comment by bardliao** [comment]
@plbossart See sof-ci/jenkins/pr-device-test, the TGLH_RVP_HDA verify-pcm-list.sh case
```
kernel: [    4.311194] sof-audio-pci 0000:00:1f.3: found 701 at link 0
kernel: [    4.311195] sof-audio-pci 0000:00:1f.3: found 701 at link 1
kernel: [    4.311197] sof-audio-pci 0000:00:1f.3: SoundWire machine driver sof_sdw topology sof-tgl-rt711-rt1308.tplg
``` 
It is a HDA device and shouldn't use sof_sdw machine driver

And the TGLU_RVP_SDW check-playback-10sec.sh case
```
kernel: [ 98.850569] rt711 sdw:0:25d:711:0: DPN_PortCtrl register write failed for port 3 
kernel: [ 98.850570] soundwire sdw-master-0: Program transport params failed: -61
kernel: [ 98.850572] soundwire sdw-master-0: Program params failed: -61 
```
and below is from the verify-pcm-list case
```
kernel: [    4.632547] rt711 sdw:0:25d:701:0: probe complete
kernel: [    4.632930] rt711 sdw:0:25d:711:0: probe complete
```
Look like rt711 codec driver is probed for both sdw:0:25d:711:0 and sdw:1:25d:700:0.

The PR do impact the existing devices. Does this PR for testing only or it is attempting to mere and upstream?

**Comment by plbossart** [comment]
Thanks @bardliao for the details, much appreciated.

This PR was an attempt to fix the HP issue, where there is no RT711 at all in the BIOS.

I do not understand at all how SoundWire could be selected on TGL-H RVP HDA. If this is an HDaudio platform, then the SoundWire link mask should be zero. Something is seriously wrong in the BIOS settings and I would assert that this PR did not change anything here.

see the dmesg trace on TGL-H:
````
kernel: [    4.272911] acpi PRP00001:00: ACPI reports 4 SDW Link devices
kernel: [    4.272918] acpi PRP00001:00: Link 2 not selected in firmware
kernel: [    4.272920] acpi PRP00001:00: Link 3 not selected in firmware
kernel: [    4.272921] acpi PRP00001:00: Creating 4 SDW Link devices
````
Link 0 and 1 should be disabled as well if HDaudio is supposed to be the configuration. This is just wrong and not something we can fix. @RanderWang @fredoh9 can you please look at BIOS settings here and the HDaudio advanced menu?

The PR has a fatal flow though in case the DSDT contains both RT711 and RT701, which is another level of brokenness. There is a hacky solution though: we can add a DMI quirk to only allow the _ADR override on specific devices and links, and report -ENODEV otherwise (it's harmless).

But first things first, why do we even have a SoundWire detection on that machine?

**Comment by plbossart** [comment]
@fredoh9 @RanderWang @bardliao I forgot to add that for TGL-SDW-RVP we would need to test without any ACPI initrd override in /lib/firmware/acpi-upgrades. If we both use the kernel workaround and the ACPI workaround, then yes the RT711 would be handled twice or never.

**Comment by RanderWang** [comment]
@aiChaoSONG  @bardliao 

> Thanks @bardliao for the details, much appreciated.
> 
> This PR was an attempt to fix the HP issue, where there is no RT711 at all in the BIOS.
> 
> I do not understand at all how SoundWire could be selected on TGL-H RVP HDA. If this is an HDaudio platform, then the SoundWire link mask should be zero. Something is seriously wrong in the BIOS settings and I would assert that this PR did not change anything here.
> 
> see the dmesg trace on TGL-H:
> 
> ```
> kernel: [    4.272911] acpi PRP00001:00: ACPI reports 4 SDW Link devices
> kernel: [    4.272918] acpi PRP00001:00: Link 2 not selected in firmware
> kernel: [    4.272920] acpi PRP00001:00: Link 3 not selected in firmware
> kernel: [    4.272921] acpi PRP00001:00: Creating 4 SDW Link devices
> ```
> 
> Link 0 and 1 should be disabled as well if HDaudio is supposed to be the configuration. This is just wrong and not something we can fix. @RanderWang @fredoh9 can you please look at BIOS settings here and the HDaudio advanced menu?
> 
> The PR has a fatal flow though in case the DSDT contains both RT711 and RT701, which is another level of brokenness. There is a hacky solution though: we can add a DMI quirk to only allow the _ADR override on specific devices and links, and report -ENODEV otherwise (it's harmless).
> 
> But first things first, why do we even have a SoundWire detection on that machine?

@aiChaoSONG  could you help to check BIOS setting on this device ? Thanks! please check CI to get device name

**Comment by bardliao** [comment]
> But first things first, why do we even have a SoundWire detection on that machine?

I am guessing HDA and SDW are not conflict in BIOS configuration. It will be a problem if we see the same configuration on real products.

**Comment by aiChaoSONG** [comment]
SOFCI TEST
@bardliao @RanderWang @plbossart On all three of our TGLH_RVP_HDA, SDW link 0 and 1 are enabled, and now all SDW links are disabled. 

>I am guessing HDA and SDW are not conflict in BIOS configuration

Yes, we can enable SDW links, SSPs, and HDA at the same time on TGLH_RVP_HDA.

**Comment by RanderWang** [comment]
> @RanderWang can you give this PR a try with no ACPI initrd override?
> 
> I was mistaken, we already have the sof-tpg-rt711-rt1308 which can work. However we are missing the DMIC autodetection, we need to figure out how to add the -2ch or -4ch suffix automagically based on NHLT information. We can't ask users to manually rename topology filenames.

@plbossart  I create a PR https://github.com/thesofproject/linux/pull/2706 to address some issues in 2700

**Review by lyakh — APPROVED** [review]
a formal LGTM ack

**Comment by plbossart** [comment]
I restricted this PR with a DMI quirk, so now there should be no need to tweak CI devices.

**Comment by RanderWang** [comment]
@plbossart this pr is good for me. I remember in early stage of sdw development you fixed such id issue in sdw layer. So I have another idea,  can we add quirk in sdw layer ? Now intel and qcom support sdw and we can create a quirk function for all platforms, just like HDA framework. Then we can fix sdw id issue at root point and we don't need to change codec driver, sof driver and machine driver.

**Comment by plbossart** [comment]
> @plbossart this pr is good for me. I remember in early stage of sdw development you fixed such id issue in sdw layer. So I have another idea, can we add quirk in sdw layer ? Now intel and qcom support sdw and we can create a quirk function for all platforms, just like HDA framework. Then we can fix sdw id issue at root point and we don't need to change codec driver, sof driver and machine driver.

I don't know how we could do this at the ACPI layer.

**Comment by plbossart** [comment]
closing, will follow @bardliao's proposal instead for a much simpler workaround solution.

**Pull request closed without merging**